plugins/repository-gcs/src/main/java/org/elasticsearch/common/blobstore/gcs/GoogleCloudStorageBlobContainer.java

```diff
@@ -0,0 +1,149 @@
+        }
+        this.keyPath = keyPath;
+    }
+
+    @Override
+    public boolean blobExists(String blobName) {
+        try {
+            return client.blobExists(bucket, buildKey(blobName));
+        } catch (IOException ioe) {
+            if (ioe instanceof GoogleJsonResponseException) {
+                GoogleJsonResponseException e = (GoogleJsonResponseException) ioe;
+                GoogleJsonError error = e.getDetails();
+                if ((e.getStatusCode() == HTTP_NOT_FOUND) || ((error != null) && (error.getCode() == HTTP_NOT_FOUND))) {
+                    return false;
+                }
+            }
+            logger.error("Failed to check if blob [{}] exists in bucket [{}]", ioe, blobName, bucket);
+            return false;
+        } catch (Throwable t) {
+            throw new BlobStoreException("Failed to check if blob [" + blobName + "] exists in bucket [" + bucket + "]", t);
+        }
+    }
+
+    @Override
+    public Map<String, BlobMetaData> listBlobs() throws IOException {
+        return listBlobsByPrefix(null);
+    }
+
+    @Override
+    public Map<String, BlobMetaData> listBlobsByPrefix(String prefix) throws IOException {
+        return client.listBlobsByPrefix(bucket, keyPath, prefix);
+    }
+
+    @Override
+    public InputStream readBlob(String blobName) throws IOException {
+        try {
+            return client.readBlob(bucket, buildKey(blobName));
+        } catch (GoogleJsonResponseException e) {
+            GoogleJsonError error = e.getDetails();
+            if ((e.getStatusCode() == HTTP_NOT_FOUND) || ((error != null) && (error.getCode() == HTTP_NOT_FOUND))) {
+                throw new FileNotFoundException(e.getMessage());
+            }
+            throw e;
+        }
+    }
+
+    @Override
+    public void writeBlob(String blobName, InputStream inputStream, long blobSize) throws IOException {
+        client.writeBlob(bucket, buildKey(blobName), inputStream, blobSize);
+    }
+
+    @Override
+    public void writeBlob(String blobName, BytesReference bytes) throws IOException {
+        client.writeBlob(bucket, buildKey(blobName), bytes.streamInput(), bytes.length());
+    }
+
+    @Override
+    public void deleteBlob(String blobName) throws IOException {
+        client.deleteBlob(bucket, buildKey(blobName));
+    }
+
+    @Override
+    public void deleteBlobsByPrefix(String prefix) throws IOException {
+        client.deleteBlobsByPrefix(bucket, buildKey(prefix));
+    }
+
+    @Override
+    public void deleteBlobs(Collection<String> blobNames) throws IOException {
+        client.deleteBlobs(bucket, buildKeys(blobNames));
+    }
+
+    @Override
+    public void move(String sourceBlobName, String targetBlobName) throws IOException {
+        client.moveBlob(bucket, buildKey(sourceBlobName), bucket, buildKey(targetBlobName));
+    }
+
+    protected String buildKey(String blobName) {
+        return keyPath + (blobName == null ? "" : blobName);
```

Comment:
I don't think we should try to work `null` here. If we got `null` here it might indicate a more serious problem and I don't think hiding it by replacing `null` with `""` is the right solutions here. I think we should just bailout here with NPE.